colossalai/shardformer/policies/t5.py

```diff
@@ -276,20 +276,27 @@ def get_held_layers(self) -> List[nn.Module]:
             if stage_manager.is_first_stage():
                 held_layers.append(model.shared)
                 held_layers.append(encoder.embed_tokens)
-                held_layers.append(encoder.dropout)
             if stage_manager.stage == decoder_starting_stage - 1:
                 held_layers.append(encoder.final_layer_norm)
                 held_layers.append(encoder.dropout)
-            held_layers.extend(encoder.block[start_idx:end_idx])
+            for i in range(start_idx, end_idx):
+                block = encoder.block[i]
+                held_layers.extend(block.layer)
+                held_layers.append(block.layer[0].SelfAttention)
```

Comment:
Why append the sub-layer alone?

Reply:
In _recusive_replace_layer in sharder.py, only modules in held_layers will be processed. If sub-layer is not appended, its sub-modules will not be replaced.

Reply:
Or maybe I should modify the logic of _recursive_replace_layer?